### DISCUSSION TITLE
使用Varchar作为主键，相同的主键的数据无法被检索

### DISCUSSION THREAD

**Discussion by yangrz7** [milvus-io/milvus#23549] [Q&A and General discussion]
```golang
func (c *MilvusClient) Insert(dim int, collectionName string, data *VectorData) (bool, error) {
	//exist, err := c.HasCollection(collectionName)
	//if err != nil {
	//	return false, err
	//}

	ctx := context.Background()

	imgIds := []string{data.ImageId}
	fields := []string{data.Field}
	stamps := []int64{data.TimeStamp}
	vectors := [][]float32{data.Vector}

	imgIdColumn := entity.NewColumnVarChar("ImgId", imgIds)
	fieldColumn := entity.NewColumnVarChar("Field", fields)
	stampColumn := entity.NewColumnInt64("TimeStamp", stamps)
	vectorColumn := entity.NewColumnFloatVector("Vector", dim, vectors)

	_, err := c.MilvusCli.Insert(ctx, collectionName, "", imgIdColumn, stampColumn, fieldColumn, vectorColumn)
	c.Log.Info("insert success")
	if err != nil {
		c.Log.Error("insert collection: ", err.Error())
		return false, err
	}

	return true, nil
}
```
创建索引相关代码
```golang
fields := []common.CollectionReq{
			{
				Name:       "ImgId",
				DataType:   21,
				AutoID:     false,
				PrimaryKey: true,
				MaxLength:  "100",
			},
			{
				Name:       "TimeStamp",
				DataType:   5,
				AutoID:     false,
				PrimaryKey: false,
			},
			{
				Name:       "Field",
				DataType:   21,
				AutoID:     false,
				PrimaryKey: false,
				MaxLength:  "100",
			},
			{
				Name:       "Vector",
				DataType:   101,
				AutoID:     false,
				PrimaryKey: false,
				VectorDims: strconv.Itoa(common.Dim),
			},
		}
```

使用ImgId(Varchar)作为主键，相同的主键的数据无法被检索，Attu中也无法显示相应的数据。
理论上应该不会覆盖才对。

**Comment by xiaofan-luan** [comment]
现在Milvus 暂不提供主键去重能力。
相同的主键，query的时候会显示后写入的。Search的时候如果新老都被搜到会只返回新的，如果只搜到老的那么只返回老的。
非常不建议写入相同的主键，这个问题暂时解决不了。如果不确定主键存不存在，可以用upsert接口（即将发布，会有额外性能开销）

**Reply by yangrz7** [reply]
我这边有个特别的需求。在某些条件下需要删除Enity, 我看到删除只能用pk。所以我用了自定义的ImgId做了主键，但是一个imgid下又可能有多个不同的图片。就导致不能全部被检索出来。这块有没有更好的方案去实现

**Reply by yhmo** [reply]
> 我这边有个特别的需求。在某些条件下需要删除Enity, 我看到删除只能用pk。所以我用了自定义的ImgId做了主键，但是一个imgid下又可能有多个不同的图片。就导致不能全部被检索出来。这块有没有更好的方案去实现

有一个迂回的办法：
1. ImgId不作为主键，而是作为一个标量字段。建表时，每个图片的主键用一个id来表示，这个id可以是auto_id，然后ImgId作为一个标量字段
2. 删除时，先用query再用delete，比如我要删除ImgId=5的所有图片:
     ids = query(expr="ImgId in [5]")
     delete(expr="id in inds")

**Reply by yangrz7** [reply]
> > 我这边有个特别的需求。在某些条件下需要删除Enity, 我看到删除只能用pk。所以我用了自定义的ImgId做了主键，但是一个imgid下又可能有多个不同的图片。就导致不能全部被检索出来。这块有没有更好的方案去实现
> 
> 有一个迂回的办法：
> 
> 1. ImgId不作为主键，而是作为一个标量字段。建表时，每个图片的主键用一个id来表示，这个id可以是auto_id，然后ImgId作为一个标量字段
> 2. 删除时，先用query再用delete，比如我要删除ImgId=5的所有图片:
>    ids = query(expr="ImgId in [5]")
>    delete(expr="id in inds")

我这边还有个问题，如果使用了自定义非自增的主键或者是相同的主键。在插入或者搜索时会产生性能问题吗 @yhmo @xiaofan-luan
